python/ray/tests/test_advanced_4.py

```diff
@@ -157,6 +161,104 @@ def ping(self):
     assert ray.get(foo.ping_actor.remote(bar)) == 3
 
 
+def get_gcs_memory_used():
+    import psutil
+    m = sum([
+        process.memory_info().rss for process in psutil.process_iter()
+        if process.name() in ("gcs_server", "redis-server")
+    ])
+    return m
+
+
+def function_entry_num(job_id):
+    from ray.ray_constants import KV_NAMESPACE_FUNCTION_TABLE
+    return len(_internal_kv_list(b"IsolatedExports:" + job_id,
+                                 namespace=KV_NAMESPACE_FUNCTION_TABLE)) + \
+        len(_internal_kv_list(b"RemoteFunction:" + job_id,
+                              namespace=KV_NAMESPACE_FUNCTION_TABLE)) + \
+        len(_internal_kv_list(b"ActorClass:" + job_id,
+                              namespace=KV_NAMESPACE_FUNCTION_TABLE))
+
+
+@pytest.mark.skipif(
+    client_test_enabled(),
+    reason="client api doesn't support namespace right now.")
+def test_function_table_gc(call_ray_start):
+    """This test tries to verify that function table is cleaned up
+    after job exits.
+    """
+
+    def f():
+        data = "0" * 1024 * 1024  # 1MB
+
+        @ray.remote
+        def r():
+            nonlocal data
+
+            @ray.remote
+            class Actor:
+                pass
+
+        return r.remote()
+
+    ray.init(address="auto", namespace="b")
+
+    # It should use > 500MB data
+    ray.get([f() for _ in range(500)])
+
+    assert get_gcs_memory_used() > 500 * 1024 * 1024
+    job_id = ray.worker.global_worker.current_job_id.binary()
+    assert function_entry_num(job_id) > 0
+    mem_size_before = get_gcs_memory_used()
+    ray.shutdown()
+
+    # We only check it for redis mode since in memory mode, the rss depends
+    # on when the OS evict the page.
+    if os.environ.get("RAY_gcs_storage") != "memory":
+        wait_for_condition(
+            lambda: (mem_size_before - get_gcs_memory_used() >= 500 * 1024 * 1024),
```

Comment:
This seems really strict, could it be flaky?

Reply:
yes, it could. maybe i should just remove this? and also in mac, it seems not working.

i think just to make sure the keys are deleted should be enough and the rest will depends on a lot of things, like db implementation and os.

Reply:
Sounds good.